### ISSUE TITLE
Bluetooth Beacons

### ISSUE THREAD

**Issue by dpalencia** [frankmaayn/app-A#57]
Beacons should detect proximity and notify user when they are near

**Comment by DarknessVolt** [comment]
https://developers.google.com/beacons/get-started

**Comment by dpalencia** [comment]
Example code
https://altbeacon.github.io/android-beacon-library/index.html

**Comment by indomie858** [comment]
Implemented beacon detection and distancing in InstructionViewActivity. When a beacon is detected, app will play 3 different beep sounds depending on the distance between beacon and phone (currently set to check 2, 5, and 10 meter distances). Checkout issue57-beacons branch to see latest updates.

Now we need to work on implementing the flow: navigate with mapbox -> switch to beacon -> you have arrived.

We also have to adjust the beacon detection so that it detects beacons with specific UUID. Our current implementation performs the distancing calcs on the first detected beacon with iBeacon protocol.

**Comment by indomie858** [comment]
Android Beacon Reference
https://github.com/AltBeacon/android-beacon-library-reference

**Comment by indomie858** [comment]
## iBeacon Parameters: UUID, Major and Minor
UUID stands for Universally Unique Identifier. It's a really long id made up of hexadecimals. It should look something like this:

> Example: f7826da6-4fa2-4e98-8024-bc5b71e0893e

Major and Minor values are used to identify beacons with greater accuracy than using UUID alone.

Minor and Major are unsigned integer values between 0 and 65535.

## UUID, Major, and Minor Configuration

#### UUID
All of our beacons should be set to share the same UUID. This is so we know the beacons are ours and are for CSUN.

#### Major
We should set different Major id values for each building.

#### Minor
We should set different Minor id values to denote entrances for each building.

**Comment by dpalencia** [comment]
Makes sense. This way we wouldn't be picking up _any_ bluetooth beacon in range; only the specific beacon we are navigating to.


Pulled up some instructions on [how to do this with a BlueCharm beacon](https://bluecharmbeacons.com/quick-start-guide-bc037-ibeacon/). Basically, you:
1. Install the EddyStone by AXAET app.
2. Open up the configuration for your bluetooth beacon from the app.
3. Set major/minor values for your beacon.

We can identify which Major/Minor ID values apply to which location by adding them as a field the Place table in the database.

**Comment by frankmaayn** [comment]
I modified my bluetooth
**For my bluetooth beacon:**
UUID: 
```
FAD17CB9C21CE3B4CD3BD3E2E80C29FE
```
MAJOR:
```
6969
```
MINOR: 
```
1738
```
NAME:
```
csunBeacon1
```
PASSWORD:
```
JK341m
```

**Comment by frankmaayn** [comment]
- [ ] Bluetooth can now detect specific UUID

Beacon documentation:
https://altbeacon.github.io/android-beacon-library/api-migration.html

**Comment by dpalencia** [comment]
Added some Major and Minor IDs to the database and some code in the InstructionVIewActivity (lines 291-312) to look for beacons by these IDs.

It doesn't quite work yet (i,.e. it still picks up any beacon and not the location-specific beacon) and I believe this is because the code executes before the ViewModel is set for the selected location, so null values are passed into the startRangingBeaconsinRegion function.

Attached is a screenshot with the ID's that I added into the database.

![Capture](https://user-images.githubusercontent.com/35319049/97849332-3fc62b00-1ca7-11eb-9292-34f5090e2807.PNG)

**Comment by dpalencia** [comment]
- Text to speech: trigger TTS notification on arrival to location; need to do this separately from IntructionVewActivity / beacon detection because beacon ranging happens every second
- Improve on beeping and vibration patterns

**Comment by dpalencia** [comment]
Note -- Camera needs a delay before being launched in InstructionViewActivity

**Comment by dpalencia** [comment]
Major/Minor - Try setting major/minor variables as members

**Comment by indomie858** [comment]
> * Text to speech: trigger TTS notification on arrival to location; need to do this separately from IntructionVewActivity / beacon detection because beacon ranging happens every second. Maybe look into doing this by triggering a service

#### Possible solution for where to implement TTS:
We could create a listener for a TextView to perform an action when the text is changed. Inside the listener, we can call whatever function is needed to execute TTS.
https://stackoverflow.com/questions/11309710/how-to-apply-the-textchange-event-on-edittext
